Update: Improve array-callback-return report message

mrflip opened the pull request:
#### What changes did you make? (Give an overview)

The array-callback-return rule should explain that a return value is required because the surrounding method requires it. This makes it clear that eg. `.filter()` expects the passed function to return something, rather than a general expectation of that function.

The error messages for expectedAtEnd/Inside now read `Method .{{arrayMethodName}}() expected a value to be returned at the end of {{name}}.` and `Method .{{arrayMethodName}}() expected a return value from {{name}}.`

#### Prerequisites checklist

- [Y] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/master/CONTRIBUTING.md).
- [Y] II signed the CLA

#### What is the purpose of this pull request? (put an "X" next to an item)

[ ] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-proposal.md))
[X] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-change-proposal.md))
[ ] Add autofixing to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

* **What rule do you want to change?**: array-callback-return
* **Does this change cause the rule to produce more or fewer warnings?**: same
* **How will the change be implemented? (New option, new default behavior, etc.)?**: new error messages
* **Please provide some example code that this change will affect:**

```js
    Object.entries(extraParams.values).map(([op, argStrs]) => {
      ops[op].extraParams = Object.fromEntries(argStrs.map((pair) => pair.split(/: */)))
    })
```

* **What does the rule currently do for this code?**: It says "Expected to return a value in arrow function". What? Who said arrow functions have to return? I do that all the time and there's no error.
* **What will the rule do after it's changed?**: It will say "Method .map() expected a return value from arrow function". Aha! I need to change .map() to .forEach().

* I Read the pull request guide (https://eslint.org/docs/developer-guide/contributing/pull-requests)
* I modified the tests for this change
* I think it automatically updates documentation for this change?

#### Is there anything you'd like reviewers to focus on?

If I were writing the test I would say 'expects' and not 'expected'; is the past tense a style guide point or an accident of history? nbd.

kaicataldo commented on lib/rules/array-callback-return.js:
```diff
@@ -153,8 +153,8 @@ module.exports = {
         messages: {
-            expectedAtEnd: "Expected to return a value at the end of {{name}}.",
-            expectedInside: "Expected to return a value in {{name}}.",
+            expectedAtEnd: "Method .{{arrayMethodName}}() expected a value to be returned at the end of {{name}}.",
```
```suggestion
            expectedAtEnd: "Array.prototype.{{arrayMethodName}}() expected a value to be returned at the end of {{name}}.",
```

kaicataldo commented on lib/rules/array-callback-return.js:
```diff
@@ -153,8 +153,8 @@ module.exports = {
-            expectedAtEnd: "Expected to return a value at the end of {{name}}.",
-            expectedInside: "Expected to return a value in {{name}}.",
+            expectedAtEnd: "Method .{{arrayMethodName}}() expected a value to be returned at the end of {{name}}.",
+            expectedInside: "Method .{{arrayMethodName}}() expected a return value from {{name}}.",
```
```suggestion
            expectedInside: "Array.prototype.{{arrayMethodName}}() expected a return value from {{name}}.",
```

kaicataldo commented:
@eslint/eslint-team Can we get a few more eyes on this? Looks like we need a champion and one more :+1: to accept this.

mdjermanovic commented:
Makes sense to me, too. I'll champion this.

If this gets accepted, could we also fix the message for `.forEach`?

Also, `Array.from` isn't `Array.prototype.from`.

> I think it automatically updates documentation for this change?

No, but I don't think there is something in the docs that should be updated for this particular change.

> If I were writing the test I would say 'expects' and not 'expected'; is the past tense a style guide point or an accident of history?

Agreed to change to `"expects"`.

mrflip commented:
I think this addresses all review points:
* Rebased to master
* forEach has a message, and describes why no return value is expected.
* Made the messages present tense ('Array.from expects no…' vs 'Array.from expected no…')
* Attributed methods correctly: `Array.from`, `Array.prototype.(everything else)`
* Added explicit check of the forEach message

@kaicataldo Could you re-review and let me know if all changes have been taken care of?

mdjermanovic commented on lib/rules/array-callback-return.js:
```diff
@@ -273,7 +273,8 @@ module.exports = {
                         data: {
-                            name: lodash.upperFirst(astUtils.getFunctionNameWithKind(funcInfo.node))
+                            name: astUtils.getFunctionNameWithKind(funcInfo.node),
+                            arrayMethodName: "Array.prototype.".concat(funcInfo.arrayMethodName)
```
This path can also report `Array.from`, e.g.:

```js
/* eslint array-callback-return: error */

// it shouldn't be "Array.prototype.from() expects a return value from arrow function"
Array.from(foo, () => { return; });
```

Maybe we could make a function that gets `funcInfo.arrayMethodName` and returns its "full name", and then use that function here and from the other `context.report`?

mrflip replied:
Thanks, those lines felt fishy and a function is the right answer. I went ahead and included all class methods (of, isArray, from) in case the function was ever lifted somewhere else. Pushed a commit with this change.

kaicataldo commented:
@eslint/eslint-team We just need one more :+1: to accept this.
